ReportCardParser/src/reportcardparser/ReportCardStats.java

```diff
@@ -42,7 +66,33 @@ public static String GetBestSubject(ReportCard reportCard) {
      * @return The average mark of the passed-in subject
      */
     public static double SubjectAverage(ReportCard[] reportCards, String subject) {
-        // REPLACE ME WITH A REAL IMPLEMENTATION!
-        return 0.0;
+        
+        //sum of all the marks of the subject
+        double marksSum = 0;
+        
+        //ittorate through all of the reportCards
+        for (ReportCard card : reportCards) {
+            
+            //add the subject's mark to the marksum
+            marksSum += getSubjectMark(card, subject).getMark();
+        }
+        
+        //return the average
+        return marksSum / reportCards.length;
+    }
+    
+    /**
+     * Get a CourseMark by name from a report card
+     * @param card the card which has the marks
+     * @param subject the name of the subject to get
+     * @return the CourseMark of the subject
+     */
+    public static CourseMark getSubjectMark(ReportCard card, String subject) {
```

Comment:
Your getSubjectMark helper method can return null, but you don’t check for this (or write a test for it).